async/gluten_async.ml

```diff
@@ -44,46 +44,35 @@ module Make_IO_Loop (Io : Gluten_async_intf.IO) = struct
     Buffer.put
       ~f:(fun buf ~off ~len k -> Async.upon (Io.read socket buf ~off ~len) k)
       read_buffer
-      (fun n -> Ivar.fill ivar n);
+      (Ivar.fill ivar);
     Ivar.read ivar
 
-  let start :
-      type t.
-      (module Gluten.RUNTIME with type t = t)
-      -> t
-      -> read_buffer_size:int
-      -> 'a fd
-      -> unit Deferred.t
+  let start (type t) m t ~read_buffer_size ~read_complete ~write_complete socket
     =
-   fun (module Runtime) t ~read_buffer_size socket ->
+    let module Runtime = (val m : Gluten.RUNTIME with type t = t) in
     let read_buffer = Buffer.create read_buffer_size in
-    let read_complete = Ivar.create () in
+    let monitor = Monitor.create () in
     let rec reader_thread () =
-      let rec reader_thread_step () =
-        match Runtime.next_read_operation t with
-        | `Read ->
-          Monitor.try_with (fun () -> read socket read_buffer) >>= ( function
-          | Ok _n ->
+      match Runtime.next_read_operation t with
+      | `Read ->
+        read socket read_buffer >>> ( function
+        | 0 ->
+          (* End of file, error, nothing can be read, exiting. *)
+          let (_ : int) = Buffer.get read_buffer ~f:(Runtime.read_eof t) in
+          Ivar.fill_if_empty read_complete ()
+        | _n ->
+          let _n =
             Buffer.get read_buffer ~f:(fun bigstring ~off ~len ->
               Runtime.read t bigstring ~off ~len)
-            |> ignore;
-            reader_thread_step ()
-          | Error End_of_file ->
-            let (_ : int) = Buffer.get read_buffer ~f:(Runtime.read_eof t) in
-            reader_thread_step ()
-          | Error exn -> raise exn )
-        | `Yield ->
-          Runtime.yield_reader t reader_thread;
-          Deferred.return ()
-        | `Close ->
-          Ivar.fill read_complete ();
-          Io.shutdown_receive socket;
-          Deferred.return ()
-      in
-      Deferred.don't_wait_for (reader_thread_step ())
+          in
+          reader_thread () )
+      | `Yield -> Runtime.yield_reader t reader_thread
+      | `Close ->
+        (* Needed ?*)
```

Comment:
I suspect this might not be needed but right now it doesn't hurt to have. I'm guessing if a program is running this branch, `read_complete` will always have been filled. Could be worth it to assert that.